### PR TITLE
Issue 5 get event data from action network

### PR DESCRIPTION
**Pull request by lperson**
Closes issue #5

### CODE REVIEW

File: lambda/action_network.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+from typing import Dict
+
+import requests
+from dateutil import parser
+
+log = logging.getLogger(__name__)
+
+
+def get_object_or_empty_dict(key: str, event: Dict) -> Dict:
+    location = event.get(key)
```
Comment (kielni) [inline lambda/action_network.py]: don't need a function to do what `.get` does: `location = event.get(key, {})` (or if key might be set to None, `event.get(key, {}) or {}`)

File: lambda/marchon.py [hunk 2]
```diff
@@ -1,22 +1,28 @@
-from datetime import datetime, timedelta
 import io
 import json
 import logging
 import os
 import traceback
+from datetime import datetime, timedelta
```
Comment (kielni) [inline lambda/marchon.py]: please leave the imports in alphabetical order

File: lambda/marchon.py [hunk 3]
```diff
@@ -76,32 +94,62 @@ def get_sheet_data():
...
+    return return_events
+
+
 def get_geojson(url):
     log.info('\nload geojson')
     resp = requests.get('https://s3.amazonaws.com/ragtag-marchon/%s' % url)
     features = {}
     for feature in resp.json()['features']:
-        features[feature['properties']['location']] = feature
+        if feature['properties'].get('source', '') == 'actionnetwork':
```
Comment (kielni) [inline lambda/marchon.py]: why is this special cased? actionnetwork events should return data in the same format

File: lambda/marchon.py [hunk 4]
```diff
@@ -76,32 +94,62 @@ def get_sheet_data():
...
+            key = feature['properties']['location']
+        features[key] = feature
     log.info('read %s features', len(features))
     return features
 
 
+def get_location_from_key(key: str) -> str:
+    parts = key.split('::', 1)
+    return parts[0]
+
```
Comment (kielni) [inline lambda/marchon.py:128]: why is this necessary?

File: lambda/marchon.py [hunk 5]
```diff
@@ -238,10 +288,19 @@ def lambda_handler(event=None, context=None, dry_run=False):
 def events_lambda_handler(event=None, context=None, dry_run=False):
     sheet = get_event_data()
     log.info('sheet=%s\n', sheet)
+
+    action_network_events = get_action_network_events()
+    log.info('action_network_events=%s\n', action_network_events)
+
+    all_events = {**sheet, **action_network_events}
```
Comment (kielni) [inline lambda/marchon.py]: this is much more cryptic and harder to read than `data.update(get_action_network_events)` where `data = get_event_data()`

File: lambda/test_action_network.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+def test_make_location_not_us():
+    inner_loc = make_test_location()
+    inner_loc['country'] = 'New Zealand'
+    location = make_location({'location': inner_loc})
+    assert location == 'New York, New Zealand'
+
+
+def test_get_object_or_empty_dict_empty_dict():
+    x = get_object_or_empty_dict('foo', {})
```
Comment (kielni) [inline lambda/test_action_network.py]: no need to write tests about python builtins

File: lambda/test_events.py [hunk 7]
```diff
@@ -1,13 +1,65 @@
...
+
+
 # print output instead of saving to S3
-marchon.events_lambda_handler(event=None, context=None, dry_run=True)
+def call_marchon_events_lambda_handler():
+    marchon.events_lambda_handler(event=None, context=None, dry_run=True)
+
+
+def test_get_geojson():
+    def response_callback(_, context):
```
Comment (kielni) [inline lambda/test_events.py]: use single quotes consistently

File: lambda/test_events.py [hunk 8]
```diff
@@ -1,13 +1,65 @@
...
 
     GOOGLE_API_KEY
     SHEET_ID
     MAPBOX_ACCESS_TOKEN
 '''
+
+
 # print output instead of saving to S3
-marchon.events_lambda_handler(event=None, context=None, dry_run=True)
+def call_marchon_events_lambda_handler():
```
Comment (kielni) [inline lambda/test_events.py]: why make a function that only calls a function? call the function directly instead

File: lambda/marchon.py [hunk 9]
```diff
@@ -76,32 +94,62 @@ def get_sheet_data():
...
+        'event': 14,
+        'eventDate': 15,
+        'eventLink': 16,
+        'photo': 17,
+        'about': 18
+    }
     return read_sheet('Sheet1!A1:S', fields, 1, True)
 
 
+def get_action_network_events() -> Dict:
```
Comment (kielni) [inline lambda/marchon.py]: why make a function that only calls a function? call the function directly instead

File: lambda/action_network.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    return next_best_email_address
+
+
+def get_email_address(event: Dict) -> str:
+    return get_email_address_from_organizer(get_organizer(event))
+
+
+def get_events_from_events_campaign(return_events=None,
+                                    page=None) -> Dict[str, Dict]:
+    if not return_events:
```
Comment (kielni) [inline lambda/action_network.py]: use the defaults in the function definition (`return_events={}, page=1`); no need to set defaults and then re-set them to different values

File: lambda/action_network.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+def make_key(converted_event: Dict) -> str:
+    return '{location}::{host}'.format(
+        location=converted_event['properties']['location'],
+        host=converted_event['properties']['host'])
+
+
+def convert_event(event: Dict) -> Dict:
+    properties = {
+        'source':
+        'actionnetwork',
```
Comment (kielni) [inline lambda/action_network.py]: put `key: value` on one line so it's easier to see what goes with what: `'source': 'actionnetwork'`

File: lambda/action_network.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+def make_location(event: Dict) -> str:
+    location = get_object_or_empty_dict('location', event)
+    if location.get('country', 'US') != 'US':
+        return '{city}, {country}'.format(
+            city=location.get('locality'), country=location.get('country'))
+    return location.get('postal_code', '')
+
+
+def get_event_name(event: Dict) -> str:
```
Comment (kielni) [inline lambda/action_network.py]: `return event.get('name', event.get('title', '')`

File: lambda/action_network.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+        if this_email_address.get('primary'):
+            return address
+        elif not next_best_email_address:
+            next_best_email_address = address
+
+    return next_best_email_address
+
+
+def get_email_address(event: Dict) -> str:
+    return get_email_address_from_organizer(get_organizer(event))
```
Comment (kielni) [inline lambda/action_network.py:54]: call it directly; don't make a function that only calls another function

File: lambda/action_network.py [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+
+
+def get_event_name(event: Dict) -> str:
+    name = event.get('name', None)
+    if not name:
+        name = event.get('title', '')
+    return name
+
+
+def get_organizer(event: Dict) -> Dict:
```
Comment (kielni) [inline lambda/action_network.py]: I really dislike one-line functions; just call it

File: lambda/action_network.py [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+    return get_object_or_empty_dict('osdi:organizer',
+                                    get_object_or_empty_dict(
+                                        '_embedded', event))
+
+
+def get_contact_name(event: Dict) -> Dict:
+    organizer = get_organizer(event)
+    given_name = organizer.get('given_name')
+    family_name = organizer.get('family_name')
+    return ' '.join([given_name, family_name])
```
Comment (kielni) [inline lambda/action_network.py]: `return '%s %s' % (organizer.get('given_name', ''), organizer.get('family_name', ''))`

need a default '' in case one of these is None; don't want 'None Smith'

File: lambda/action_network.py [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+        parser.parse(event.get('start_date',
+                               '1/20/2018')).strftime('%-m/%-d/%Y'),
+        'eventLink':
+        event.get('browser_url', ''),
+        'location':
+        make_location(event),
+        'contactEmail':
+        get_email_address(event),
+        'host':
+        get_contact_name(event),
```
Comment (kielni) [inline lambda/action_network.py]: call this once and reuse the return value

File: lambda/action_network.py [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+        return {}
+
+    log.info('\nsuccess! get_events_from events_campaign (action network) -- page %d',
+             page)
+
+    response_json = response.json()
+    events = response_json.get('_embedded', {}).get('osdi:events', {})
+    for event in events:
+        converted_event = convert_event(event)
+        return_events[make_key(converted_event)] = converted_event
```
Comment (kielni) [inline lambda/action_network.py]: maybe `convert_event` could return `{ 'xyz': { 'properties': ... }` and then you can just do `return_events.update(convert_event(event))`

File: lambda/action_network_sample_data.py [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+            'osdi:taggings': {
+                'href':
+                'https://actionnetwork.org/api/v2/people/38cc1710-4a71-4225-9607-e8269758e051/taggings'
+            },
+            'self': {
+                'href':
+                'https://actionnetwork.org/api/v2/people/38cc1710-4a71-4225-9607-e8269758e051'
+            }
+        },
+        'created_date':
```
Comment (kielni) [inline lambda/action_network_sample_data.py]: key simple key-value pairs on the same line

**Review by kielni — CHANGES_REQUESTED** [review]

File: lambda/action_network.py (hunk 10 above)
Comment (lperson) [reply lambda/action_network.py]: Using {} as a default argument is a bad idea.  The next time the function is called, it will still have items populated from the current invocation.  Agree about page.

File: lambda/marchon.py (hunk 5 above)
Comment (lperson) [reply lambda/marchon.py]: I disagree.

File: lambda/action_network.py (hunk 12 above)
Comment (lperson) [reply lambda/action_network.py]: I did the extra check because this will not return title if name is in the dictionary with the value  '' or None.  

(This is also why I write tests that seemingly test built in functionality.  I'd rather test it explicitly than rely on my memory of language details like this.)

```
>>> d = {'a':None}
>>> print(d.get('a', 'b'))
None
```

In the above example, this works.  It returns b.

```
d.get('a') or 'b'
```

A small function is reusable, easier to read, and less error prone.

I did get rid of the get_event_name and instead used 

```
'name': event.get('name') or event.get('title') or '',
```

File: lambda/action_network.py (hunk 15 above)
Comment (lperson) [reply lambda/action_network.py]: I would agree, put the '' default in get.  We get space-Smith with either your code or mine with a default of '', so we may want to do something like:

I think assigning to organizer, given_name and family_name improves readability, and I'm willing to give up the conciseness.

```
name=''
if given_name:
    name = given_name
if given_name and family_name:
    name += ' '
if family_name
    name += family_name
```

Comment (lperson) [reply lambda/action_network.py]: I prefer the reality to the conciseness.

File: lambda/action_network.py (hunk 14 above)
Comment (lperson) [reply lambda/action_network.py]: get_organizer may be a one_line function but there's a lot of complexity in there that I think is worth encapsulating.

File: lambda/action_network.py (hunk 1 above)
Comment (lperson) [reply lambda/action_network.py]: This is called 3 times.  I feel that the function call is cleaner than typing event.get(key) or {}) 3 times.  I'll get rid of it, though.

File: lambda/marchon.py (hunk 2 above)
Comment (lperson) [reply lambda/marchon.py]: isort did that; I'll change it back; I believe it is correct according to python coding standards the way I have it.

File: lambda/marchon.py (hunk 4 above)
Comment (lperson) [reply lambda/marchon.py:128]: 1. You said zip code is a sufficient location for house parties on this map
2. You said zip code works reliably for geocoding in map box
3. To point 2, `New York, NY 10025` failed to geocode while `NY, NY 10025` did geocode, so I went with zip code only because it's reliable and accurate enough
4. There could be more than one party in a given zip code

File: lambda/test_action_network.py (hunk 6 above)
Comment (lperson) [reply lambda/test_action_network.py]: get_object_or_empty_dict was not a builtin

and as I said above, I feel more confident seeing an explicit test than I do relying on my memory of the intricacies of the behavior of object.get

File: lambda/action_network.py (hunk 17 above)
Comment (lperson) [reply lambda/action_network.py]: Interesting ...

File: lambda/action_network.py (hunk 11 above)
Comment (lperson) [reply lambda/action_network.py]: yapf is doing this.  I'll change it manually and disable yapf around it.

File: lambda/marchon.py (hunk 3 above)
Comment (lperson) [reply lambda/marchon.py]: Explained below.  This allows more than one action network event in the same location.

**Comment by lperson** [comment]
@kielni Thanks for the thoughtful code review!

File: lambda/marchon.py (hunk 3 above)
Comment (lperson) [reply lambda/marchon.py]: @kielni perhaps one way to address this is to get rid of the special casing and use the same logic (allowing multiple events per location) for all events.

File: lambda/marchon.py (hunk 5 above)
Comment (lperson) [reply lambda/marchon.py]: I thought about this.  I agree now.  Someone looking at the code for the first time may be confused by this.

Comment (lperson) [reply lambda/marchon.py]: Confession: I was excited about the python 3.5 syntax.

File: lambda/marchon.py (hunk 3 above)
Comment (kielni) [reply lambda/marchon.py]: ok that makes sense; maybe add a comment in the code so we can remember why they're different later without having to dig into another file?

File: lambda/test_events.py [hunk 19]
```diff
@@ -1,13 +1,63 @@
...
-# print output instead of saving to S3
-marchon.events_lambda_handler(event=None, context=None, dry_run=True)
+
+
+
+
+def test_get_geojson():
+    def response_callback(_, context):
+        features = {
+            "type":
```
Comment (kielni) [inline lambda/test_events.py]: use single quotes consistently
